File: logic-engineering/scripts/compile_iir.mjs
#!/usr/bin/env node
import {readJson,writeJson,rootOf,buildNodeIndex,semanticHash} from './lib/model.mjs';

function pascal(value){return String(value??'Generated').split(/[^A-Za-z0-9]+/).filter(Boolean).map(x=>x[0].toUpperCase()+x.slice(1)).join('')||'Generated';}
function camel(value){const p=pascal(value);return p[0].toLowerCase()+p.slice(1);}
function entityOfField(ref){const parts=String(ref).split('.');return parts.length>=3?parts.slice(0,2).join('.'):null;}
function collectRefs(v,out=new Set()){if(Array.isArray(v))for(const x of v)collectRefs(x,out);else if(v&&typeof v==='object'){if(typeof v.ref==='string')out.add(v.ref);for(const x of Object.values(v))collectRefs(x,out);}return out;}

function compileDomain(root){
  const enums=[],entities=[],runtime=[];
  for(const node of root.domain??[]){
    if(node.kind==='enum') enums.push({semantic_ref:node.id,name:pascal(node.id.replace(/^domain\./,'')),display_name:node.name??null,values:[...(node.values??[])]});
    if(node.kind==='entity'){
      const slot=camel(node.id.replace(/^domain\./,''));
      const fields=(node.fields??[]).map(f=>({semantic_ref:f.id,name:String(f.id).split('.').at(-1),type_ref:f.type,nullable:f.nullable??false,cardinality:f.cardinality??'one'}));
      entities.push({semantic_ref:node.id,name:pascal(node.id.replace(/^domain\./,'')),display_name:node.name??null,slot,fields});
      runtime.push({semantic_ref:node.id,kind:'entity',slot});
      for(const f of fields) runtime.push({semantic_ref:f.semantic_ref,kind:'field',entity_ref:node.id,slot,path:[f.name]});
    }
  }
  return {domain_types:{enums,entities},runtime_bindings:runtime};
}

function repositoryId(resource){const entity=entityOfField(resource)??resource;return `repository.${String(entity??'resource').replace(/^domain\./,'').replace(/\./g,'_')}`;}
function compileContracts(root){
  const repoMap=new Map(),portMap=new Map();
  for(const e of root.effects??[]){
    if(['read','write','persist'].includes(e.kind)){
      const entityRef=entityOfField(e.resource)??e.resource??'domain.unknown';
      const id=repositoryId(e.resource);
      if(!repoMap.has(id))repoMap.set(id,{id,kind:'repository_contract',entity_ref:entityRef,semantic_refs:[],operations:[],binding:{strategy:'repository',provider:null}});
      const repo=repoMap.get(id);repo.semantic_refs.push(e.id);const name=e.kind==='read'?'load':'save';if(!repo.operations.some(x=>x.name===name))repo.operations.push({name,semantic_refs:[e.id]});
    }else if(['external_call','emit','schedule'].includes(e.kind)){
      const system=e.system??(e.kind==='emit'?'event_bus':e.kind);
      const id=`port.${String(system).replace(/[^A-Za-z0-9_.-]/g,'_')}`;
      if(!portMap.has(id))portMap.set(id,{id,kind:'external_port',system,semantic_refs:[],operations:[],generation_mode:'contract_only'});
      const port=portMap.get(id);port.semantic_refs.push(e.id);const name=e.operation??(e.kind==='emit'?'publish':'execute');if(!port.operations.some(x=>x.name===name))port.operations.push({name,semantic_refs:[e.id]});
    }
  }
  return {repositories:[...repoMap.values()],ports:[...portMap.values()]};
}

function compilePlans(root,profile,unresolved){
  const transaction_plans=[],concurrency_plans=[],retry_plans=[],idempotency_plans=[];
  for(const c of root.constraints??[]){
    if(c.kind==='atomicity'){
      const strategy=profile.transaction_strategy??null;
      transaction_plans.push({id:`plan.${c.id}`,kind:'transaction_plan',semantic_refs:[c.id],members:c.actions??c.members??[],strategy,provider:profile.persistence??null});
      if(!strategy)unresolved.push({semantic_ref:c.id,reason:'Target Profile 缺少 transaction_strategy',required_for:c.id,severity:'blocking'});
    }else if(c.kind==='concurrency'){
      const strategy=profile.concurrency_strategy??null;
      concurrency_plans.push({id:`plan.${c.id}`,kind:'concurrency_plan',semantic_refs:[c.id],resource_ref:c.resource??null,scope:c.scope??null,strategy});
      if(!strategy)unresolved.push({semantic_ref:c.id,reason:'Target Profile 缺少 concurrency_strategy',required_for:c.id,severity:'blocking'});
    }else if(c.kind==='idempotency'){
      const strategy=profile.idempotency_strategy??c.strategy??null;
      idempotency_plans.push({id:`plan.${c.id}`,kind:'idempotency_plan',semantic_refs:[c.id],operation_ref:c.operation??null,key_ref:c.key??null,strategy});
      if(!strategy)unresolved.push({semantic_ref:c.id,reason:'缺少幂等实现策略',required_for:c.id,severity:'blocking'});
    }else if(c.kind==='retry'){
      const strategy=profile.retry_strategy??c.strategy??null;
      retry_plans.push({id:`plan.${c.id}`,kind:'retry_plan',semantic_refs:[c.id],operation_ref:c.operation??null,strategy,max_attempts:c.max_attempts??null});
      if(!strategy)unresolved.push({semantic_ref:c.id,reason:'缺少重试实现策略',required_for:c.id,severity:'blocking'});
    }
  }
  return {transaction_plans,concurrency_plans,retry_plans,idempotency_plans};
}

function compileUseCases(root,index,repositories,ports,unresolved){
  const use_cases=[];
  for(const b of root.behaviors??[]){
    const guards=[];const steps=[];const refs=new Set();
    for(const id of b.preconditions??[]){const n=index.get(id);if(!n){unresolved.push({semantic_ref:id,reason:'Behavior precondition 节点不存在',required_for:b.id,severity:'blocking'});continue;}guards.push({semantic_ref:id,expression:n.expression??null,failure_ref:n.failure??null});collectRefs(n.expression,refs);}
    for(const id of b.flow??[]){const n=index.get(id);if(!n){unresolved.push({semantic_ref:id,reason:'Behavior flow 节点不存在',required_for:b.id,severity:'blocking'});continue;}const step={semantic_ref:id,kind:n.kind,operation:n.operation??null,target:n.target??null,value:n.value??null,effects:n.effects??[],when:n.when??null,then:n.then??[],else:n.else??[]};steps.push(step);collectRefs(step,refs);}
    for(const post of b.postconditions??[]){const n=index.get(post);if(n?.expression)collectRefs(n.expression,refs);}
    const inputEntities=[...new Set([...refs].map(entityOfField).filter(Boolean))];
    const effectIds=new Set(steps.flatMap(s=>s.effects??[]));
    const dependencies=[...repositories.filter(x=>x.semantic_refs.some(r=>effectIds.has(r))).map(x=>x.id),...ports.filter(x=>x.semantic_refs.some(r=>effectIds.has(r))).map(x=>x.id)];
    use_cases.push({id:`usecase.${b.id.replace(/^behavior\./,'')}`,kind:'use_case',semantic_refs:[b.id],name:pascal(b.id.replace(/^behavior\./,'')),display_name:b.name??null,input_refs:inputEntities,inputs:inputEntities,guards,steps,outputs:b.outputs??[],failure_refs:b.failures??[],postconditions:b.postconditions??[],dependencies});
  }
  return use_cases;
}

function compileErrors(root){
  const refs=new Set();for(const b of root.behaviors??[])for(const x of b.failures??[])refs.add(x);for(const r of root.rules??[])if(r.failure)refs.add(r.failure);
  return [...refs].map(ref=>({id:`error_mapping.${ref.replace(/^error\./,'')}`,semantic_error_ref:ref,target_error:`${pascal(ref.replace(/^error\./,''))}Error`}));
}

function compile(clm,profile){
  const root=rootOf(clm),index=buildNodeIndex(clm),unresolved=[];
  const {domain_types,runtime_bindings}=compileDomain(root);
  const {repositories,ports}=compileContracts(root);
  for(const repo of repositories){repo.binding.provider=profile.persistence??null;if(!profile.persistence)unresolved.push({semantic_ref:repo.entity_ref,reason:'Target Profile 缺少 persistence provider',required_for:repo.id,severity:'blocking'});}
  const plans=compilePlans(root,profile,unresolved);
  const use_cases=compileUseCases(root,index,repositories,ports,unresolved);
  const error_mappings=compileErrors(root);
  const generation_regions=[...use_cases.map(u=>({id:`region.${u.id}`,mode:'generated',semantic_refs:u.semantic_refs})),...repositories.map(r=>({id:`region.${r.id}`,mode:'contract_only',semantic_refs:r.semantic_refs.length?r.semantic_refs:[r.entity_ref]})),...ports.map(p=>({id:`region.${p.id}`,mode:p.generation_mode,semantic_refs:p.semantic_refs}))];
  const traceability=use_cases.map(u=>({implementation_id:u.id,semantic_refs:[...u.semantic_refs,...u.guards.map(g=>g.semantic_ref),...u.steps.map(s=>s.semantic_ref)]}));
  return {iir:{version:'0.2',source_clm_id:root.id,source_clm_version:root.version,source_semantic_hash:semanticHash(clm),target_profile:profile,domain_types,runtime_bindings,use_cases,repository_contracts:repositories,external_ports:ports,...plans,error_mappings,primitive_bindings:[],generation_regions,traceability,unresolved}};
}

const [clmFile,profileFile,...args]=process.argv.slice(2);const oi=args.indexOf('-o')>=0?args.indexOf('-o'):args.indexOf('--output');const output=oi>=0?args[oi+1]:null;
if(!clmFile||!profileFile||!output){console.error('usage: node compile_iir.mjs clm.json target-profile.json -o implementation.iir.json');process.exit(2);}
try{const clm=readJson(clmFile),raw=readJson(profileFile),profile=raw.target_profile??raw;writeJson(output,compile(clm,profile));console.log(JSON.stringify({ok:true,output},null,2));}catch(e){console.error(JSON.stringify({ok:false,error:e.message},null,2));process.exit(1);}

File: logic-engineering/scripts/compile_iir.py
#!/usr/bin/env python3
"""将 CLM 编译为最小实现中间表示（IIR）。

第一版只做确定性技术映射：
- Behavior → use case
- Rule → guard / validation
- Action → operation
- Effect → persistence / external IO / event
- Constraint → transaction / concurrency / idempotency requirement
- Primitive → target binding

节点索引统一使用 clm_model.py，避免 rules / decisions / actions 被遗漏。
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any, Dict

from clm_model import build_node_index, root_of


def load_json(path: Path) -> Dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def save_json(path: Path, value: Dict[str, Any]) -> None:
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def compile_target_profile(profile: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "language": profile.get("language"),
        "version": profile.get("version"),
        "framework": profile.get("framework"),
        "architecture": profile.get("architecture"),
        "persistence": profile.get("persistence"),
        "messaging": profile.get("messaging"),
        "transaction_strategy": profile.get("transaction_strategy"),
        "error_model": profile.get("error_model"),
        "test_framework": profile.get("test_framework"),
    }


def compile_rule(rule: Dict[str, Any]) -> Dict[str, Any]:
    condition: Dict[str, Any]
    if "expression" in rule:
        condition = {"expression": rule.get("expression")}
    elif "conditions" in rule:
        condition = {"operator": rule.get("operator"), "conditions": rule.get("conditions", [])}
    else:
        condition = {
            "operator": rule.get("operator"),
            "subject": rule.get("subject"),
            "value": rule.get("value"),
        }
    return {
        "semantic_id": rule.get("id"),
        "kind": "guard",
        "condition": condition,
        "failure": rule.get("failure"),
    }


def compile_action(action: Dict[str, Any], idx: Dict[str, Dict[str, Any]]) -> Dict[str, Any]:
    result = {
        "semantic_id": action.get("id"),
        "kind": action.get("kind", "action"),
        "operation": action.get("operation"),
        "target": action.get("target"),
        "value": action.get("value"),
        "effects": [],
    }
    for ref in action.get("effects", []) or []:
        effect = idx.get(ref)
        if effect:
            result["effects"].append({
                "semantic_id": ref,
                "kind": effect.get("kind"),
                "resource": effect.get("resource"),
                "system": effect.get("system"),
                "operation": effect.get("operation"),
                "idempotency": effect.get("idempotency"),
            })
        else:
            result["effects"].append({"semantic_id": ref, "unresolved": True})
    if action.get("kind") == "foreach":
        result.update({
            "collection": action.get("collection"),
            "item": action.get("item"),
            "when": action.get("when"),
            "do": action.get("do", []),
        })
    return result


def compile_decision(decision: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "semantic_id": decision.get("id"),
        "kind": "decision",
        "when": decision.get("when"),
        "then": decision.get("then", []),
        "else": decision.get("else", []),
    }


def compile_constraint(c: Dict[str, Any]) -> Dict[str, Any]:
    result = {"semantic_id": c.get("id"), "kind": c.get("kind")}
    for key in (
        "expression", "trigger", "requirement", "time_bound", "resource",
        "scope", "policy", "actions", "guarantee", "operation", "key",
    ):
        if key in c:
            result[key] = c[key]
    return result


def compile_behavior(behavior: Dict[str, Any], idx: Dict[str, Dict[str, Any]]) -> Dict[str, Any]:
    guards = []
    for ref in behavior.get("preconditions", []) or []:
        rule = idx.get(ref)
        guards.append(compile_rule(rule) if rule else {"semantic_id": ref, "unresolved": True})

    steps = []
    for ref in behavior.get("flow", []) or []:
        node = idx.get(ref)
        if not node:
            steps.append({"semantic_id": ref, "unresolved": True})
        elif node.get("kind") == "decision":
            steps.append(compile_decision(node))
        else:
            steps.append(compile_action(node, idx))

    return {
        "semantic_id": behavior.get("id"),
        "name": behavior.get("name"),
        "kind": "use_case",
        "inputs": behavior.get("inputs", []),
        "guards": guards,
        "steps": steps,
        "outputs": behavior.get("outputs", []),
        "failures": behavior.get("failures", []),
        "postconditions": behavior.get("postconditions", []),
    }


def compile_primitive_bindings(root: Dict[str, Any], profile: Dict[str, Any]) -> list[Dict[str, Any]]:
    profile_id = profile.get("id") or profile.get("name")
    bindings = []
    for p in root.get("primitives", []) or []:
        raw = p.get("bindings", {}) or {}
        binding = None
        if profile_id and profile_id in raw:
            binding = raw[profile_id]
        elif profile.get("binding_key") and profile["binding_key"] in raw:
            binding = raw[profile["binding_key"]]
        bindings.append({
            "semantic_id": p.get("id"),
            "binding": binding,
            "required_contract": p.get("contract", {}),
            "resolved": binding is not None,
        })
    return bindings


def compile_iir(clm: Dict[str, Any], profile: Dict[str, Any]) -> Dict[str, Any]:
    root = root_of(clm)
    idx = build_node_index(root)
    iir = {
        "iir": {
            "source_clm_id": root.get("id"),
            "source_clm_version": root.get("version"),
            "target_profile": compile_target_profile(profile),
            "use_cases": [],
            "constraints": [],
            "state_models": [],
            "primitive_bindings": [],
            "unresolved": [],
        }
    }
    out = iir["iir"]

    for b in root.get("behaviors", []) or []:
        out["use_cases"].append(compile_behavior(b, idx))

    for c in root.get("constraints", []) or []:
        out["constraints"].append(compile_constraint(c))

    for s in root.get("states", []) or []:
        if s.get("kind") == "state_machine":
            transitions = []
            for ref in s.get("transitions", []) or []:
                t = idx.get(ref)
                if t:
                    transitions.append({
                        "semantic_id": t.get("id"),
                        "from": t.get("from"),
                        "to": t.get("to"),
                        "trigger": t.get("trigger"),
                        "guard": t.get("guard"),
                    })
                else:
                    transitions.append({"semantic_id": ref, "unresolved": True})
            out["state_models"].append({
                "semantic_id": s.get("id"),
                "states": s.get("states", []),
                "transitions": transitions,
            })

    out["primitive_bindings"] = compile_primitive_bindings(root, profile)

    for use_case in out["use_cases"]:
        for guard in use_case["guards"]:
            if guard.get("unresolved"):
                out["unresolved"].append(guard["semantic_id"])
        for step in use_case["steps"]:
            if step.get("unresolved"):
                out["unresolved"].append(step["semantic_id"])
            for effect in step.get("effects", []) or []:
                if effect.get("unresolved"):
                    out["unresolved"].append(effect["semantic_id"])
    for binding in out["primitive_bindings"]:
        if not binding["resolved"]:
            out["unresolved"].append(binding["semantic_id"])

    out["unresolved"] = sorted(set(out["unresolved"]))
    return iir


def main() -> int:
    parser = argparse.ArgumentParser(description="CLM → IIR 最小编译器")
    parser.add_argument("clm", type=Path)
    parser.add_argument("profile", type=Path)
    parser.add_argument("-o", "--output", type=Path)
    args = parser.parse_args()

    try:
        clm = load_json(args.clm)
        profile = load_json(args.profile)
        profile = profile.get("target_profile", profile)
        iir = compile_iir(clm, profile)
    except (OSError, json.JSONDecodeError) as exc:
        print(json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False, indent=2))
        return 1

    if args.output:
        save_json(args.output, iir)
    else:
        print(json.dumps(iir, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
